### PR TITLE
[Executor] Support exec_line_async executor APIs

### PR DESCRIPTION
**Pull request by Jasmin3q**
# Description

Support exec_line_async executor APIs

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 1]
```diff
@@ -753,6 +805,34 @@ def _get_node_referenced_flow_inputs(
                 node_referenced_flow_inputs[value.value] = flow_inputs[value.value]
         return node_referenced_flow_inputs
 
+    def _exec_prepare(self, inputs, run_id, line_number, variant_id):
```
Comment (guming-learning) [inline src/promptflow/promptflow/executor/flow_executor.py:808]: The function returns a 4-elements tuple, and it seems very hard to think of a proper function name, maybe we can keep duplicated code for now.

@thy09 what do you think?

Comment (thy09) [reply src/promptflow/promptflow/executor/flow_executor.py:808]: Can we use `with` to achieve this?
This ask is not that hurry so we can take more time to improve these.

Comment (thy09) [reply src/promptflow/promptflow/executor/flow_executor.py:808]: Another point, since variant_id is no longer used, we can remove this param.

**Review by guming-learning — DISMISSED** [review]

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 2]
```diff
@@ -973,7 +1112,24 @@ def _traverse_nodes(self, inputs, context: FlowExecutionContext) -> Tuple[dict,
...
+            flow_logger.info("Start executing nodes in thread pool mode.")
+            nodes_outputs, bypassed_nodes = self._submit_to_scheduler(context, inputs, batch_nodes)
+        outputs = self._extract_outputs(nodes_outputs, bypassed_nodes, inputs)
+        return outputs, nodes_outputs
+
+    async def _traverse_nodes_async(self, inputs, context: FlowExecutionContext) -> Tuple[dict, dict]:
+        batch_nodes = [node for node in self._flow.nodes if not node.aggregation]
+        outputs = {}
+        #  TODO: Use a mixed scheduler to support both async and thread pool mode.
+        if self._should_use_async():
```
Comment (thy09) [inline src/promptflow/promptflow/executor/flow_executor.py:1129]: When calling from async function, always use async schedueler.

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 3]
```diff
@@ -973,7 +1112,24 @@ def _traverse_nodes(self, inputs, context: FlowExecutionContext) -> Tuple[dict,
         if self._should_use_async():
             flow_logger.info("Start executing nodes in async mode.")
             scheduler = AsyncNodesScheduler(self._tools_manager, self._node_concurrency)
-            nodes_outputs, bypassed_nodes = asyncio.run(scheduler.execute(batch_nodes, inputs, context))
+            # Use async_run_allow_running_loop to avoid calling asyncio.run from a loop when running a notebook.
+            nodes_outputs, bypassed_nodes = async_run_allowing_running_loop(
```
Comment (thy09) [inline src/promptflow/promptflow/executor/flow_executor.py:1116]: I don't think we have such scenario, why do we modify this?

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 4]
```diff
@@ -906,6 +958,93 @@ def _exec(
...
+            flow_id=self._flow_id,
+            line_number=line_number,
+            variant_id=variant_id,
+        )
+        output = {}
+        aggregation_inputs = {}
+        try:
+            if validate_inputs:
+                inputs = FlowValidator.ensure_flow_inputs_type(flow=self._flow, inputs=inputs, idx=line_number)
+            inputs = load_multimedia_data(self._flow.inputs, inputs)
```
Comment (thy09) [inline src/promptflow/promptflow/executor/flow_executor.py:1017]: Add some TODOs here.
load_multimedia_data/stringify_generator_output/persist/end_run may requires async implementation

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 3 above)
Comment (Jasmin3q) [reply src/promptflow/promptflow/executor/flow_executor.py:1116]: There are test cases that run this code in a jupyter notebook. If we don't use async_run_allowing_running_loop, those cases will fail.

**Review by thy09 — APPROVED** [review]

**Review by crazygao — APPROVED** [review]

**Pull request merged**